engine/lib/filestore.php

```diff
@@ -598,6 +578,230 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
+ * @param array $segments URL segments
+ * @return void
+ * @access private
+ * @since 2.2
+ */
+function _elgg_filestore_embed_file_handler($segments) {
+
+	// clear cache-boosting headers set by PHP session
+	header_remove('Cache-Control');
+	header_remove('Pragma');
+	header_remove('Expires');
+
+	$request = _elgg_services()->request;
+	$response = new Symfony\Component\HttpFoundation\Response();
+	$response->prepare($request);
+
+	$guid = array_shift($segments);
+	$size = array_shift($segments);
+
+	$file = get_entity($guid);
+	if (!$file instanceof ElggFile) {
+		if (elgg_entity_exists($guid)) {
+			$response->setStatusCode(403)->setContent('You are not allowed to access this file')->send();
+		} else {
+			$response->setStatusCode(404)->setContent('File does not exist')->send();
+		}
+		exit;
+	}
+
+	$thumbnail = elgg_get_thumbnail($file, $size);
+	if (!$thumbnail->exists()) {
+		$response->setStatusCode(404)->setContent('Thumbnail doest not exist')->send();
+		exit;
+	}
+
+	$if_none_match = $request->headers->get('if_none_match');
+	if (!empty($if_none_match)) {
+		// strip mod_deflate suffixes
+		$request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+	}
+
+	$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+	$last_updated = filemtime($filenameonfilestore);
+	$etag = '"' . $last_updated . '"';
+	$response->setPublic()->setEtag($etag);
+	if ($response->isNotModified($request)) {
+		$response->send();
+		exit();
+	}
+
+	$headers = [
+		'Content-Type' => (new Elgg\Filesystem\MimeTypeDetector())->getType($filenameonfilestore),
+	];
+	$response = new Symfony\Component\HttpFoundation\BinaryFileResponse($filenameonfilestore, 200, $headers, false, 'inline');
+	$response->prepare($request);
+
+	$expires_dt = (new DateTime())->setTimestamp(strtotime('+1 day'));
+	$response->setExpires($expires_dt);
+
+	$response->setEtag($etag);
+	
+	$response->send();
+	exit;
+}
+
+/**
+ * Returns thumbnail sizes configuration for a file entity
+ *
+ * @param \ElggFile $file File entity
+ * @return array
+ */
+function elgg_get_thumbnail_sizes(\ElggFile $file) {
+
+	$sizes = [
+		'small' => [
+			'w' => 60,
+			'h' => 60,
+			'square' => true,
+			'metadata_name' => 'thumbnail',
+			'filename_prefix' => 'thumb',
+		],
+		'medium' => [
+			'w' => 153,
+			'h' => 153,
+			'square' => true,
+			'metadata_name' => 'smallthumb',
+			'filename_prefix' => 'smallthumb',
+		],
+		'large' => [
+			'w' => 600,
+			'h' => 600,
+			'square' => false,
+			'metadata_name' => 'largethumb',
+			'filename_prefix' => 'largethumb',
+		],
+	];
+
+	// @todo: add a hook to filter these values
+	return $sizes;
+}
+
+/**
+ * Attempt to create thumbnails for a file
+ * For non image files (or on a failed attempt to resize source image), all previously generated thumbnails will be cleared
+ *
+ * @param \ElggFile $file File entity
+ * @return bool
+ */
+function elgg_create_thumbnails(\ElggFile $file) {
+
+	if (!$file->guid) {
+		return false;
+	}
+
```

Comment:
I was tempted to add ElggFile::getSimpleType() and patch up
ElggFile::getMimeType(). I will do it while at it.
On Apr 4, 2016 5:06 PM, "Steve Clay" <notifications@github.com> wrote:

> In engine/lib/filestore.php
> <https://github.com/Elgg/Elgg/pull/9608#discussion_r58390883>:
>
> > +	return $sizes;
> > +}
> > +
> > +/**
> > + * Attempt to create thumbnails for a file
> > + * For non image files (or on a failed attempt to resize source image), all previously generated thumbnails will be cleared
> > + *
> > + * @param \ElggFile $file File entity
> > + * @return bool
> > + */
> > +function elgg_create_thumbnails(\ElggFile $file) {
> > +
> > +	if (!$file->guid) {
> > +		return false;
> > +	}
> > +
>
> Alternately we could make simpletype an argument. That way callers know it
> has to be set first.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9608/files/3dc274ba889edb909d01ca478946274a72d0baf2#r58390883>
>